service/src/java/org/apache/hive/service/server/KillQueryImpl.java

```diff
@@ -36,18 +54,78 @@ public KillQueryImpl(OperationManager operationManager) {
     this.operationManager = operationManager;
   }
 
+  public static Set<ApplicationId> getChildYarnJobs(Configuration conf, String tag) {
+    Set<ApplicationId> childYarnJobs = new HashSet<ApplicationId>();
+    GetApplicationsRequest gar = GetApplicationsRequest.newInstance();
+    gar.setScope(ApplicationsRequestScope.OWN);
+    gar.setApplicationTags(Collections.singleton(tag));
+
+    try {
+      ApplicationClientProtocol proxy = ClientRMProxy.createRMProxy(conf, ApplicationClientProtocol.class);
+      GetApplicationsResponse apps = proxy.getApplications(gar);
+      List<ApplicationReport> appsList = apps.getApplicationList();
+      for(ApplicationReport appReport : appsList) {
+        childYarnJobs.add(appReport.getApplicationId());
+      }
+    } catch (YarnException | IOException ioe) {
+      throw new RuntimeException("Exception occurred while finding child jobs", ioe);
+    }
+
+    if (childYarnJobs.isEmpty()) {
+      LOG.info("No child applications found");
+    } else {
+      LOG.info("Found child YARN applications: " + StringUtils.join(childYarnJobs, ","));
+    }
+
+    return childYarnJobs;
+  }
+
+  public static void killChildYarnJobs(Configuration conf, String tag) {
+    try {
+      Set<ApplicationId> childYarnJobs = getChildYarnJobs(conf, tag);
+      if (!childYarnJobs.isEmpty()) {
+        YarnClient yarnClient = YarnClient.createYarnClient();
+        yarnClient.init(conf);
+        yarnClient.start();
+        for (ApplicationId app : childYarnJobs) {
+          yarnClient.killApplication(app);
+        }
+      }
+    } catch (IOException | YarnException ye) {
+      throw new RuntimeException("Exception occurred while killing child job(s)", ye);
+    }
+  }
+
   @Override
   public void killQuery(String queryId, String errMsg) throws HiveException {
     try {
+      String queryTag = null;
       Operation operation = operationManager.getOperationByQueryId(queryId);
       if (operation == null) {
-        LOG.info("Query not found: " + queryId);
+        // Check if user has passed the query tag to kill the operation.
+        operation = operationManager.getOperationByQueryTag(queryId);
+        if (operation == null) {
+          LOG.info("Query not found: " + queryId);
+        }
       } else {
+        queryTag = operation.getQueryTag();
+      }
+
+      if (operation != null) {
```

Comment:
Better to invoke cancelOperation after killing the yarn jobs to avoid losing tagToIdMap entry if anything fails in this method.

Reply:
done